### PR TITLE
support gradle init and wrapper

### PR DESCRIPTION
**Pull request by atomfrede**
This PR add initial gradle support for init and wrapper. I decided to use kotlin dsl. 

The PR is still in draft mode as I just wanted to know if the structure is fine or if something totally wrong and should be changed. After your go @pascalgrimaud I will go on and add needles for dependencies and so on.

updates #408

### CODE REVIEW

**Comment by pascalgrimaud** [comment]
Oh thanks a lot @atomfrede 
Just have a very very quick review and the structure seems good
I can review it tomorrow but I'd prefer adding needles, dependencies etc in separated PRs (using this PR as start point) if possible. It will be easier to review, if you're ok

**Comment by atomfrede** [comment]
Sure even better. Take your time no need to hurry. Will need some testing if the kotlin dsl syntax is complete

**Comment by atomfrede** [comment]
@pascalgrimaud The first part is done and everythings works fine. I did not check the sonar duplications yet, but as maven and gradle are quite similar I expected that. I would not try find and extract common parts to be honest.

File: src/main/java/tech/jhipster/lite/generator/buildtool/gradle/domain/Gradle.java [hunk 1]
```diff
@@ -0,0 +1,4 @@
+package tech.jhipster.lite.generator.buildtool.gradle.domain;
+
+public class Gradle {
+}
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/buildtool/gradle/domain/Gradle.java:4]:
```suggestion
import java.util.Map;

public class Gradle {

  private Gradle() {}

  public static Map<String, String> gradleWrapper() {
    // prettier-ignore
    return Map.of(
      "gradlew", ".",
      "gradlew.bat", ".",
      "gradle-wrapper.jar", "gradle/wrapper",
      "gradle-wrapper.properties", "gradle/wrapper"
    );
  }
}
```

File: src/main/java/tech/jhipster/lite/generator/buildtool/gradle/domain/GradleDomainService.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    projectRepository.add(project, SOURCE, "gradlew");
+    projectRepository.setExecutable(project, "", "gradlew");
+
+    projectRepository.add(project, SOURCE, "gradlew.bat");
+    projectRepository.setExecutable(project, "", "gradlew.bat");
+
+    String sourceWrapper = getPath(SOURCE, "gradle", "wrapper");
+    String destinationWrapper = getPath("gradle", "wrapper");
+    projectRepository.add(project, sourceWrapper, "gradle-wrapper.jar", destinationWrapper);
+    projectRepository.add(project, sourceWrapper, "gradle-wrapper.properties", destinationWrapper);
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/buildtool/gradle/domain/GradleDomainService.java:57]:
```suggestion
    Gradle
      .gradleWrapper()
      .forEach((file, location) -> projectRepository.add(project, getPath(SOURCE, location), file, location));
    projectRepository.setExecutable(project, "", "gradlew");
    projectRepository.setExecutable(project, "", "gradlew.bat");
```

File: src/main/java/tech/jhipster/lite/generator/buildtool/gradle/README.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# Maven
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/buildtool/gradle/README.md:1]:
```suggestion
# Gradle
```

File: src/main/java/tech/jhipster/lite/generator/buildtool/gradle/application/GradleApplicationService.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  public GradleApplicationService(GradleService gradleService) {
+    this.gradleService = gradleService;
+  }
+
+  public void init(Project project) {
+    gradleService.initJava(project);
+  }
+
+  public void addBuildGradle(Project project) {
+    gradleService.addJavaBuildGradleKts(project);
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/buildtool/gradle/application/GradleApplicationService.java:21]: not covered by int test

File: src/main/java/tech/jhipster/lite/generator/buildtool/gradle/infrastructure/primary/rest/GradleResource.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  private final GradleApplicationService gradleApplicationService;
+
+  public GradleResource(GradleApplicationService gradleApplicationService) {
+    this.gradleApplicationService = gradleApplicationService;
+  }
+
+  @Operation(summary = "Init", description = "Init Gradle project with build.gradle.kts and wrapper")
+  @ApiResponse(responseCode = "500", description = "An error occurred while initializing Gradle project")
+  @PostMapping
+  @GeneratorStep(id = GeneratorAction.MAVEN_JAVA)
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/buildtool/gradle/infrastructure/primary/rest/GradleResource.java:30]: need a new constant here

File: src/main/java/tech/jhipster/lite/generator/buildtool/gradle/infrastructure/primary/rest/GradleResource.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  public void init(@RequestBody ProjectDTO projectDTO) {
+    Project project = ProjectDTO.toProject(projectDTO);
+    gradleApplicationService.init(project);
+  }
+
+  @Operation(summary = "Add build.gradle.kts")
+  @ApiResponse(responseCode = "500", description = "An error occurred while adding build.gradle.kts to project")
+  @PostMapping("/build-gradle")
+  @GeneratorStep(id = GeneratorAction.MAVEN_JAVA_POM_XML)
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/buildtool/gradle/infrastructure/primary/rest/GradleResource.java:39]: same, need a new constant here

File: src/main/java/tech/jhipster/lite/generator/buildtool/gradle/infrastructure/primary/rest/GradleResource.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  public void addPomXml(@RequestBody ProjectDTO projectDTO) {
+    Project project = ProjectDTO.toProject(projectDTO);
+    gradleApplicationService.addBuildGradle(project);
+  }
+
+  @Operation(summary = "Add Gradle Wrapper")
+  @ApiResponse(responseCode = "500", description = "An error occurred while adding Gradle Wrapper to project")
+  @PostMapping("/wrapper")
+  @GeneratorStep(id = GeneratorAction.MAVEN_WRAPPER)
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/buildtool/gradle/infrastructure/primary/rest/GradleResource.java:48]: new constant too

File: src/main/java/tech/jhipster/lite/generator/buildtool/gradle/domain/GradleDomainService.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  @Override
+  public Optional<String> getVersion(String name) {
+    return Optional.empty();
+  }
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/buildtool/gradle/domain/GradleDomainService.java:63]: if not needed yet, you can remove I think

File: src/main/java/tech/jhipster/lite/generator/buildtool/gradle/domain/GradleService.java [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+import java.util.Optional;
+
+public interface GradleService {
+
+  void initJava(Project project);
+
+  void addJavaBuildGradleKts(Project project);
+  void addGradleWrapper(Project project);
+
+  Optional<String> getVersion(String name);
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/buildtool/gradle/domain/GradleService.java:14]: remove if it's not needed yet

File: src/main/java/tech/jhipster/lite/generator/buildtool/gradle/infrastructure/primary/rest/GradleResource.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+  public void init(@RequestBody ProjectDTO projectDTO) {
+    Project project = ProjectDTO.toProject(projectDTO);
+    gradleApplicationService.init(project);
+  }
+
+  @Operation(summary = "Add build.gradle.kts")
+  @ApiResponse(responseCode = "500", description = "An error occurred while adding build.gradle.kts to project")
+  @PostMapping("/build-gradle")
+  @GeneratorStep(id = GeneratorAction.MAVEN_JAVA_POM_XML)
+  public void addPomXml(@RequestBody ProjectDTO projectDTO) {
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/buildtool/gradle/infrastructure/primary/rest/GradleResource.java:40]: rename method

File: src/main/java/tech/jhipster/lite/generator/buildtool/gradle/infrastructure/primary/rest/GradleResource.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    Project project = ProjectDTO.toProject(projectDTO);
+    gradleApplicationService.addBuildGradle(project);
+  }
+
+  @Operation(summary = "Add Gradle Wrapper")
+  @ApiResponse(responseCode = "500", description = "An error occurred while adding Gradle Wrapper to project")
+  @PostMapping("/wrapper")
+  @GeneratorStep(id = GeneratorAction.MAVEN_WRAPPER)
+  public void addMavenWrapper(@RequestBody ProjectDTO projectDTO) {
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/buildtool/gradle/infrastructure/primary/rest/GradleResource.java:49]: rename method

File: src/main/java/tech/jhipster/lite/generator/buildtool/gradle/infrastructure/primary/rest/GradleResource.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    gradleApplicationService.init(project);
+  }
+
+  @Operation(summary = "Add build.gradle.kts")
+  @ApiResponse(responseCode = "500", description = "An error occurred while adding build.gradle.kts to project")
+  @PostMapping("/build-gradle")
+  @GeneratorStep(id = GeneratorAction.MAVEN_JAVA_POM_XML)
+  public void addPomXml(@RequestBody ProjectDTO projectDTO) {
+    Project project = ProjectDTO.toProject(projectDTO);
+    gradleApplicationService.addBuildGradle(project);
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/buildtool/gradle/infrastructure/primary/rest/GradleResource.java:42]: not covered by test

File: src/main/resources/generator/buildtool/gradle/build.gradle.kts.mustache [hunk 13]
```diff
@@ -0,0 +1,56 @@
+plugins {
```
Comment (pascalgrimaud) [inline src/main/resources/generator/buildtool/gradle/build.gradle.kts.mustache:1]: wonder if we should format with 2 spaces for indent, to be consistent with all the appliaction

File: src/main/resources/generator/buildtool/gradle/build.gradle.kts.mustache [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+repositories {
+    mavenCentral()
+    // jhipster-needle-gradle-repository
+}
+
+group = "{{packageName}}{{^packageName}}com.mycompany.myapp{{/packageName}}"
+version = "0.0.1-SNAPSHOT"
+
+ext{
+  set("junit-jupiter.version", "5.8.2")
```
Comment (pascalgrimaud) [inline src/main/resources/generator/buildtool/gradle/build.gradle.kts.mustache:22]: the indentation is good here :)

File: src/test/java/tech/jhipster/lite/generator/buildtool/gradle/infrastructure/primary/rest/GradleResourceIT.java [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+
+    Project project = ProjectDTO.toProject(projectDTO);
+    assertFilesGradle(project);
+
+    assertFileContent(project, SETTINGS_GRADLE_KTS, "rootProject.name = \"chips\"");
+    assertFileContent(project, BUILD_GRADLE_KTS,  "group = \"tech.jhipster.chips\"");
+  }
+
+  @Test
+  void shouldAddMavenWrapper() throws Exception {
```
Comment (pascalgrimaud) [inline src/test/java/tech/jhipster/lite/generator/buildtool/gradle/infrastructure/primary/rest/GradleResourceIT.java:48]: rename test

**Review by pascalgrimaud — CHANGES_REQUESTED** [review]

**Comment by atomfrede** [comment]
Thanks for the review! Will have a look later today.

**Comment by pascalgrimaud** [comment]
For maven, I launch: ./mvnw clean verify
What is the similar command for gradle plz?

**Comment by atomfrede** [comment]
./gradlew check should do it

File: src/main/java/tech/jhipster/lite/generator/buildtool/gradle/domain/GradleDomainService.java (hunk 8 above)
Comment (atomfrede) [reply src/main/java/tech/jhipster/lite/generator/buildtool/gradle/domain/GradleDomainService.java:63]: Still thinking how to get the version.

File: src/main/resources/generator/buildtool/gradle/build.gradle.kts.mustache (hunk 13 above)
Comment (atomfrede) [reply src/main/resources/generator/buildtool/gradle/build.gradle.kts.mustache:1]: Will do, we should

File: src/main/java/tech/jhipster/lite/generator/buildtool/gradle/domain/GradleDomainService.java (hunk 8 above)
Comment (atomfrede) [reply src/main/java/tech/jhipster/lite/generator/buildtool/gradle/domain/GradleDomainService.java:57]: Remove for now as not needed (yet)

**Review by pascalgrimaud — APPROVED** [review]

**Comment by pascalgrimaud** [comment]
Thanks and good job @atomfrede 
I'll take care of the CI, as it needs some customisation before adding Gradle there :)

**Comment by pascalgrimaud** [comment]
Strange, I got this error:

```
22:38:07 in /tmp/jhlite/gradleapp on  main took 4s 
➜ ./gradlew check
> Task :test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> No tests found for given includes: [**/*Test.*](filter.includeTestsMatching)

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 3s
3 actionable tasks: 1 executed, 2 up-to-date

22:38:15 in /tmp/jhlite/gradleapp on  main took 3s 
➜ ll src/test/java/tech/jhipster/gradleapp/error/domain/*Test.*
-rw-rw-r-- 1 pgrimaud pgrimaud 4,4K avril  9 22:35 src/test/java/tech/jhipster/gradleapp/error/domain/AssertTest.java
-rw-rw-r-- 1 pgrimaud pgrimaud 1,3K avril  9 22:35 src/test/java/tech/jhipster/gradleapp/error/domain/MissingMandatoryValueExceptionTest.java
-rw-rw-r-- 1 pgrimaud pgrimaud 2,9K avril  9 22:35 src/test/java/tech/jhipster/gradleapp/error/domain/UnauthorizedValueExceptionTest.java

```

**Comment by atomfrede** [comment]
Ah maybe I just did compile when testing it. But yes when there are no tests the tasks fails (at least I had that with another project already). Let me check tomorrow if it can be configured.

**Comment by pascalgrimaud** [comment]
@atomfrede : in order, I use: init, gradle, java-base, so there are some tests (see my logs). I didn't find the solution yet, as I'm noob with Gradle :-D

**Comment by atomfrede** [comment]
That's really strange. Let me check. Maybe something with the kotlin dsl ( first time using it)

**Comment by pascalgrimaud** [comment]
I found this https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/testing/TestFilter.html and using this, seems better:

image

image

**Comment by atomfrede** [comment]
Thanks @pascalgrimaud ! Indeed difference between groovy and kotlin dsl as the other exclusion works fine with the groovy dsl

**Comment by atomfrede** [comment]
Bounty claimed https://opencollective.com/generator-jhipster/expenses/71921

**Comment by pascalgrimaud** [comment]
@atomfrede : welcome to this project and thanks for your work here :)
Approved!

**Pull request merged**